src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -31,14 +31,12 @@ internal class SslStreamInternal
         private int _nestedRead;
         private AsyncProtocolRequest _readProtocolRequest; // cached, reusable AsyncProtocolRequest used for read operations
         private AsyncProtocolRequest _writeProtocolRequest; // cached, reusable AsyncProtocolRequest used for write operations
+        private Action<Task, object> _freeBufferAction;
```

Comment:
So here is the thing (and I very well could be wrong, maybe you can comment?) the way I read the code, there is overlapping IO. In that you can write to the underlying socket, while that is going on you have already scheduled the next "chunk". This means that you aren't using a pooled buffer for the overlapping encryption chunk. In my testing (using console.writeline) I could see that the current code doesn't use the pooled arrays but every second for larger writes. 

If that assumption is correct. Then using a member variable only ever let us track a single buffer in flight. By using a state object in an action like I have done we can take a second buffer if needed to overlap, once I did that there was no need to allocate on the overlap call.